### PR TITLE
Add a toplevel stanza

### PR DESCRIPTION
**Pull request by rgrinberg**
As discussed in the retreat, a toplevel stanza is useful.

A couple of points:

* It's possible to improve this feature further with subsystems. This might be useful to give things like mdx and our toplevel tests first class support.

* There are some `buildable` flags that make sense on the `toplevel` stanza that haven't been included yet. We can include those flags now or perhaps add them later when there's a real need.

### CODE REVIEW

**Review by emillon — APPROVED** [review]

**Comment by avsm** [comment]
Just so I understand this feature's scope, is it a solution to https://github.com/ocaml/dune/issues/688? I.e. will the `toplevel` stanza generate a `.top` subpackage that installs printers for the toplevel reliably?

**Comment by avsm** [comment]
That makes sense! I do think it would be useful to automatically and systematically expose a way to have top packages, and load then automatically into dune utop. But that is indeed a different feature.

**Review by ghost — APPROVED** [review]

**Comment by rgrinberg** [comment]
Okay, I've added some tests for this feature and will shortly add documentation. Since the feature seems to be taken well, I'd like to ask for some feedback on how to extend this feature in a next PR to support user defined/custom toplevels.

What I'd like is a way for users to create their own custom toplevel libraries. The idea being to implement things such as our `expect_test.mll` library. I see a couple of possibilities:

* We could let the user specify a `main` field where they can write the function that will start the toplevel. They'll need to specify the toplevel library in the list of libraries. It's a bit annoying to have to specify this if you're defining several toplevels of the same type however.

* We could have the user specify a `toplevel_library` that would be responsible for initializing the toplevel. I'm wondering if this approach will be tricky from the point of side effects. I.e. we want to make sure that `Clflags.include` is loaded before the library executes.

Other suggestions are welcome.

**Comment by diml** [comment]
The first option seems better to me, and in practice it's unlikely users will define more than one toplevel. Another option would be to make `toplevel` behave a bit like `exeuctable`. I.e. you wouldn't need a separate library for the entry point of the toplevel, however that would make it harder to define both a library and a toplevel in the same directory.

BTW, given that we have `dune utop`, should we make it easy to declare custom utop toplevels? For instance by adding a `(utop)` field to the `toplevel` stanza?

File: src/toplevel.ml [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+  let main_module_name t = Module.Name.of_string t.name
+  let main_module_filename t = t.name ^ ".ml"
+  let source_path t = Path.relative t.dir (main_module_filename t)
+
+  let obj_dir { dir; name ; _ } =
+    Obj_dir.make_exe ~dir ~name
+
+  let exe t =
+    (* Use the [.exe] version. As the utop executable is declared with [(modes
```
Comment (diml) [inline src/toplevel.ml]: There are a few mention to utop in this file that should probably be replaced by something more generic.

Comment (rgrinberg) [reply src/toplevel.ml]: Done. This function wasn't used so I just deleted it.

**Comment by rgrinberg** [comment]
> BTW, given that we have dune utop, should we make it easy to declare custom utop toplevels? For instance by adding a (utop) field to the toplevel stanza?

Would this be mutually exclusive with setting `main`? We could also have a `kind` parameter like this:

```
type kind =
  | Vanila
  | Utop
  | Custom of string (* main *)
```

**Comment by diml** [comment]
It would yeah. The `kind` parameter seems good.

**Comment by rgrinberg** [comment]
Docs & Issue to track the issue above are added.

**Pull request merged**